engine/server/engine/enclave_manager/enclave_pool.go

```diff
@@ -203,34 +242,58 @@ func (pool *EnclavePool) destroyPreviousIdleEnclaves() error {
 }
 
 // run method is in charge of filling the enclave pool channel with new idle enclaves
-func (pool *EnclavePool) run() {
+func (pool *EnclavePool) run(ctx context.Context) {
 
-	defer close(pool.idleEnclavesChan)
+	defer func() {
+		close(pool.idleEnclavesChan)
+	}()
 
-	fails := 0
+	// init the pool
+	if err := pool.fill(ctx); err != nil {
+		logrus.Errorf("An error occurred filling the enclave pool. Error\n%v", err)
+		return
+	}
 
 	for {
-		newEnclaveInfo, err := pool.createNewEnclave()
-		if err != nil {
-			// Retry strategy if something fails
-			logrus.Errorf("An error occurred creating a new idle enclave. Error\n%v", err)
-			fails++
-			if fails >= maxRetryAfterFails {
-				logrus.Debugf("The enclave pool sub-routine has been canceled after reaching the max retries value '%v'", maxRetryAfterFails)
+		// wait until receive the re-fill signal or the close signal
+		select {
+		case <-pool.fillChan:
+			if err := pool.fill(ctx); err != nil {
+				if err == context.Canceled {
+					logrus.Debug("The sub-routine context has been canceled")
+				} else {
+					logrus.Errorf("An error occurred filling the enclave pool. Error\n%v", err)
+				}
 				break
 			}
-			logrus.Debugf("Next retry will be executed on '%v' seconds", timeBetweenRetries.Seconds())
-			time.Sleep(timeBetweenRetries)
+		case <-pool.closeChan:
```

Comment:
Can't you pass in the cancelCtx and replace this with `case <-cancelCtx.Done()` ? This way you don't even need the closeChan, as soon as you call the `cancelFunc` in your Close() function, it will send a signal here

Reply:
replaced! good catch!